core/lib/zksync_core/src/vm_runner/output_handler.rs

```diff
@@ -0,0 +1,204 @@
+use std::{
+    fmt::{Debug, Formatter},
+    sync::Arc,
+    time::Duration,
+};
+
+use async_trait::async_trait;
+use dashmap::DashMap;
+use futures::future::BoxFuture;
+use tokio::sync::{oneshot, watch};
+use zksync_dal::Core;
+use zksync_db_connection::connection_pool::ConnectionPool;
+use zksync_types::L1BatchNumber;
+
+use crate::{
+    state_keeper::{updates::UpdatesManager, StateKeeperOutputHandler},
+    vm_runner::VmRunnerStorageLoader,
+};
+
+#[async_trait]
+pub trait OutputHandlerFactory: Debug + Send {
+    async fn create_handler(
+        &mut self,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<Box<dyn StateKeeperOutputHandler>>;
+}
+
+pub struct ConcurrentOutputHandlerFactory<L: VmRunnerStorageLoader, F: OutputHandlerFactory> {
+    pool: ConnectionPool<Core>,
+    state: Arc<DashMap<L1BatchNumber, oneshot::Receiver<BoxFuture<'static, anyhow::Result<()>>>>>,
+    loader: L,
+    factory: F,
```

Comment:
Do you think it's better to keep factory here instead of `StateKeeperOutputHandler`? From the top of my head I don't see a use-case where factory is required, do you have any? Maybe I miss something and factory is required here then please explain

Reply:
Hmm, so my thinking was that each output handler instance would be short-lived. Meaning it gets created when the user asks for a new handler and then gets destroyed when `handle_l1_batch` is called. Thus every `AsyncOutputHandler` would own its own internal handler that gets created through the factory and is destroyed once `AsyncOutputHandler` is dropped.

The other option would be to share one underlying handler for all `AsyncOutputHandler`s, but then we would have to wrap it in `Arc<RwLock<...>>` and then acquire the lock every time we call `handle_l2_block()` etc which would create thread competition where it is not strictly needed.